Update tensorify pass to specialize symfloats we didn't tensorify away

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138868

As discussed w/ @ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

bobrenjc93 committed:
Update tensorify pass to specialize symfloats we didn't tensorify away

[ghstack-poisoned]

ezyang commented on test/inductor/test_torchinductor_dynamic_shapes.py:
```diff
@@ -972,15 +972,17 @@ def test_unspecialized_float_operations(self):
+                z = 1.3
+                self.assertEqual(fn(x, 2.0, z), fn_opt(x, 2.0, z))
+                self.assertEqual(fn(x, 3.0, z), fn_opt(x, 3.0, z))
                 self.assertEqual(cnt.frame_count, 1)
```
It's up to your discretion, but you potentially might want to add this as a separate test, rather than overwriting the old one.

ezyang commented on torch/fx/node.py:
```diff
@@ -78,6 +78,7 @@
 _side_effectful_functions: Set[Callable] = {
     torch._assert,
     torch._assert_async,
+    _ops.aten.copy_.default,
```
you shouldn't need to do this. Specifically, this operator should have a schema that says it's mutable and so should automatically be considered side effectful.

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -102,24 +106,35 @@ def tensorify_python_scalars(
     placeholders = set()
     for node in graph.nodes:
         if node.op != "placeholder":
+            first_none_placeholder = node
```
s/none/non/ ?

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -176,25 +191,24 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
         ):
             # Look for tensor.item() calls on placeholders
-            if unbacked_bindings := node.meta.get("unbacked_bindings"):
-                for s in unbacked_bindings.keys():
```
good

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +260,72 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
                     node.replace_all_uses_with(replacement_proxy.node)
-
                     graph.erase_node(node)
+                    deleted.add(node)
```
Not sure why you need to track these, because you've already erased them from the graph

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -102,24 +106,35 @@ def tensorify_python_scalars(
 
     Analysis = TensorReferenceAnalysis
 
+    def dce(graph: fx.Graph, deleted: set[fx.Node]) -> None:
```
Not sure why this is implemented from scratch as opposed to using eliminate_dead_code with a custom predicate.

bobrenjc93 replied:
Refactored in the next diff. As discussed in chat, custom predicate doesn't eliminate item calls as we need but the default predicate does: https://github.com/pytorch/pytorch/pull/138869

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +260,72 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+
+    # Now do one more pass that specializes all symfloats we didn't manage
+    # to tensorify away.
+    nodes = list(graph.nodes)
```
from docs:

> Note that this Node list representation is a doubly-linked list. Mutations during iteration (e.g. delete a Node, add a Node) are safe.

ezyang replied:
So it shouldn't be necessary to make a copy of the node list

bobrenjc93 replied:
I need it for `graph.inserting_before` call

```
        with graph.inserting_before(
            nodes[i + 1] if node not in placeholders else first_non_placeholder
        ):
```

Or perhaps I'm misunderstanding. Can you propose a suggested change if so?

ezyang replied:
ok appeal to preexisting condition accepted lol

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +260,72 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+            if (
+                callable(node.target)
+                and len(args) > 0
+                and all(isinstance(arg, (float)) for arg in args)
```
nit: no parens around float

bobrenjc93 replied:
Is there an easy way to add a rule so lintrunner will fix this for me automatically?

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +260,72 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+    dce(graph, deleted)
+
+    # Now do one more pass that specializes all symfloats we didn't manage
+    # to tensorify away.
```
I think you can write this pass differently, that avoids the need to do another DCE at the end. The idea is to do the pass is REVERSE, rather than front-to-back.

When doing the pass in reverse, we look for operators that have backed SymFloat arguments, where we have a hint for it (maybe we also want to check if the operator supported it? Not sure, my primary fear is that we are going to obliterate something like math.floor(math.sqrt(x.size(0)), I want these to stay in the graph and not specialize them. Maybe a better test is something like, check that there aren't any free regular integer symbols in the expression). When this occurs, we issue a guard (you also have a choice here, whether or to guard on the expression, or guard on each of the free unbacked symfloats individually; the latter is less risky but the former preserves more dynamism.) And then you can immediately swap the node.

And then if you like, you can also DCE as you go, because if you kill all the users of a symfloat expression then things will start become eliminatable as you go backwards.

Will be good to write some tests for this.

Does this make sense?

bobrenjc93 replied:
Can you elaborate on this case?

```
math.floor(math.sqrt(x.size(0))
```

`x.size(0)` returns a SymInt not a SymFloat right? We only specialize away SymFloats so this shouldn't be a concern?

Regarding writing the pass in reverse to eliminate a DCE pass, that does make sense. Given how hard these diffs have been to get green and this pass technically does work correctly, I'll try and stack a refactor diff on top of this one that does the pass in reverse.

ezyang replied:
I'm talking about in my hypothetical rewrite. you will process the output of math.floor first, which potentially is eligible for specialization.

ezyang replied:
I don't mind you stacking on top but if we're not in a rush I will just wait to land all of them together

bobrenjc93 replied:
The only benefit of landing these earlier is we can start kicking the tires and seeing what breaks in the wild. I don't have strong opinions either way but generally I like to err on the side of merging these semantically correct changes earlier and then landing subsequent refactors. The time delay is also useful to discern if breakages are actually due to the refactor itself.

ezyang replied:
This is a reasonable argument, but my experience in pytorch has taught me that landing closely interrelated changes at too close of an interval (a few days, which is the case here since you got CI working on all of them) together is more pain than gain. The problem is that if the early PR *does* cause a problem not only does that PR have to be backed out, but all of the ones that depend on it on top. Sure you did find out that it was the original PR and not the refactor that caused the problem... but if there's a minimal repro you would have easily been able to figure that out anyway.

I'm sympathetic to keeping the refactor PR stacked on top for CI signal (because maybe you did break something in the refactor that was fine in the original), and I'm also sympathetic to going ahead and landing the original PR if it's all ready to go. But you don't even have ciflow/trunk on this PR (so you can't land it right now), AND I see you are clean CI on the two PRs stacked on top, as of Sunday evening. Might as well squash them together.

bobrenjc93 replied:
Sounds good. I think you were one step ahead of me. FWIW I didn't anticipate being able to get the CI green so quickly :p.

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +247,84 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+        if node.op == "output" or node.op == "placeholder":
+            continue
+
+        with graph.inserting_before(nodes[i]):
```
Isn't this just 'node'?

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +247,84 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+            for a in node.args:
+                if (
+                    isinstance(a, fx.Node)
+                    and (zf := a.meta.get("val")) is not None
```
nit: zf naming here is a bit misleading, since typically the zf here won't be a symbol, it will be a generic sympy expression

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +247,84 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+                if (
+                    isinstance(a, fx.Node)
+                    and (zf := a.meta.get("val")) is not None
+                    and isinstance(zf, (torch.SymFloat, torch.SymInt))
```
For completeness, probably should have SymBool here too

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +247,84 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+                    isinstance(a, fx.Node)
+                    and (zf := a.meta.get("val")) is not None
+                    and isinstance(zf, (torch.SymFloat, torch.SymInt))
+                    and (node_hint := zf.node.hint) is not None
```
nit: walrus is dead

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +247,84 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+                    # The high level idea here is to specialize if we have
+                    # a node that has exclusively SymFloat free symbols.
+                    # This is because we don't want to specialize cases like
+                    # math.floor(math.sqrt(x.size(0))).
```
Here is a near miss. Suppose we have `some_op(math.sqrt(x.size(0)) + input_float)`.  Here, there is a dependence on both `s0` and on `zf0`. Let's say we were not able to tensorify `some_op`. We would like to force `input_float` to be specialized. However, because there is an `s0` free symbol here, we don't guard on this expression (and rightly so.)

Will `input_float` eventually get specialized later in the pass? It's not clear that this happens. The problem is that the pass here only inspects args/kwargs of operators it's trying to specialize. So if `input_float` is literally the input argument, you will never force spec, which seems bad.

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +247,42 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+                (val := node.meta.get("val")),
+                (torch.SymFloat, torch.SymInt, torch.SymBool),
+            ):
+                if len(val.node.expr.free_symbols) > 0 and all(
```
what happens when there are no free symbols :0 is it already constant then?

ezyang replied:
It feels like it's safe to run the code below anyway

bobrenjc93 replied:
Yeah as discussed offline there's an underlying bug in bool specialization. I'll fix the bug and remove this condition. In particular this test `python test/dynamo/test_dynamic_shapes.py DynamicShapesMiscTests.test_runtime_assert_replacement_dynamic_shapes` 

without this condition results in the voided asserts without proper guarding

graph: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/bobren/ab43b2c1-6932-4de3-98d5-8773931aaa3a/custom/0_0_0/inductor_post_grad_graph_4.txt
guards: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/bobren/ab43b2c1-6932-4de3-98d5-8773931aaa3a/custom/0_0_0/dynamo_cpp_guards_str_7.txt

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -244,13 +247,42 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
+                    elif isinstance(val, torch.SymInt):
+                        node.replace_all_uses_with(int(val))
+                    elif isinstance(val, torch.SymBool):
+                        node.replace_all_uses_with(bool(val))
```
Nit you can use `guard_scalar` here from torch.fx.experimental.symbolic_shapes

ezyang reviewed (approved):
very slick

bobrenjc93 committed:
Update on "Update tensorify pass to specialize symfloats we didn't tensorify away"


As discussed w/ ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge -f "ci failures unrelated"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but I think the new tests are failing on fbcode' -c ghfirst

The diff is D65247988 and the failure is:

```
======================================================================
ERROR: test_unspecialized_float_fallback_specialization_cuda (caffe2.test.inductor.test_torchinductor_dynamic_shapes.TestInductorDynamicCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/testing/_internal/common_utils.py", line 3052, in wrapper
    method(*args, **kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/testing/_internal/common_device_type.py", line 480, in instantiated_test
    raise rte
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/testing/_internal/common_device_type.py", line 460, in instantiated_test
    result = test(self, **param_kwargs)
  File "/usr/local/fbcode/platform010/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/caffe2/test/inductor/test_torchinductor_dynamic_shapes.py", line 1005, in test_unspecialized_float_fallback_specialization
    self.assertEqual(fn(x, 2.0, z), fn_opt(x, 2.0, z))
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/eval_frame.py", line 554, in _fn
    return fn(*args, **kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/convert_frame.py", line 1428, in __call__
    return self._torchdynamo_orig_callable(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/convert_frame.py", line 1211, in __call__
    result = self._inner_convert(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/convert_frame.py", line 548, in __call__
    return _compile(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/convert_frame.py", line 981, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/convert_frame.py", line 707, in compile_inner
    return _compile_inner(code, one_graph, hooks, transform)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_utils_internal.py", line 350, in wrapper_function
    result = StrobelightCompileTimeProfiler.profile_compile_time(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/caffe2/fb/strobelight/compile_time_profiler.py", line 162, in profile_compile_time
    return func(*args, **kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/convert_frame.py", line 742, in _compile_inner
    out_code = transform_code_object(code, transform)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/bytecode_transformation.py", line 1337, in transform_code_object
    transformations(instructions, code_options)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/convert_frame.py", line 232, in _fn
    return fn(*args, **kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/convert_frame.py", line 661, in transform
    tracer.run()
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/symbolic_convert.py", line 2909, in run
    super().run()
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/symbolic_convert.py", line 1115, in run
    while self.step():
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/symbolic_convert.py", line 1027, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/symbolic_convert.py", line 3100, in RETURN_VALUE
    self._return(inst)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/symbolic_convert.py", line 3085, in _return
    self.output.compile_subgraph(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/output_graph.py", line 1173, in compile_subgraph
    self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/output_graph.py", line 1411, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/output_graph.py", line 1458, in call_user_compiler
    return self._call_user_compiler(gm)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/output_graph.py", line 1507, in _call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/output_graph.py", line 1488, in _call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/repro/after_dynamo.py", line 130, in __call__
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/testing.py", line 233, in __call__
    return lookup_backend(self.backend)(gm, example_inputs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/backends/inductor.py", line 12, in inductor
    return compile_fx(*args, **kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/compile_fx.py", line 1693, in compile_fx
    return aot_autograd(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/backends/common.py", line 72, in __call__
    cg = aot_module_simplified(gm, example_inputs, **self.kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_functorch/aot_autograd.py", line 1105, in aot_module_simplified
    compiled_fn = dispatch_and_compile()
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_functorch/aot_autograd.py", line 1081, in dispatch_and_compile
    compiled_fn, _ = create_aot_dispatcher_function(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_functorch/aot_autograd.py", line 528, in create_aot_dispatcher_function
    return _create_aot_dispatcher_function(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_functorch/aot_autograd.py", line 780, in _create_aot_dispatcher_function
    compiled_fn, fw_metadata = compiler_fn(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 196, in aot_dispatch_base
    compiled_fw = compiler(fw_module, updated_flat_args)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/compile_fx.py", line 1510, in fw_compiler_base
    return _fw_compiler_base(model, example_inputs, is_inference)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/compile_fx.py", line 1579, in _fw_compiler_base
    return inner_compile(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/compile_fx.py", line 578, in compile_fx_inner
    return wrap_compiler_debug(_compile_fx_inner, compiler_name="inductor")(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_dynamo/repro/after_aot.py", line 100, in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/fb/utils.py", line 167, in newFunction
    return old_func(*args, **kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/compile_fx.py", line 735, in _compile_fx_inner
    compiled_graph = FxGraphCache.load(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/codecache.py", line 1460, in load
    compiled_graph = compile_fx_fn(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/compile_fx.py", line 642, in codegen_and_compile
    compiled_graph = fx_codegen_and_compile(gm, example_inputs, **fx_kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/compile_fx.py", line 934, in fx_codegen_and_compile
    graph.run(*example_inputs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/graph.py", line 820, in run
    return super().run(*args)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/fx/interpreter.py", line 167, in run
    self.env[node] = self.run_node(node)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/graph.py", line 1411, in run_node
    result = super().run_node(n)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/fx/interpreter.py", line 228, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/graph.py", line 1060, in call_function
    raise LoweringException(e, target, args, kwargs).with_traceback(
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/graph.py", line 1057, in call_function
    out = lowerings[target](*args, **kwargs)  # type: ignore[index]
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/lowering.py", line 401, in wrapped
    out = decomp_fn(*args, **kwargs)
  File "/dev/shm/uid-30083/48acfd58-seed-nspid4026537968_cgpid14780696-ns-4026537918/torch/_inductor/lowering.py", line 2841, in _local_scalar_dense
    V.graph.sizevars.shape_env, V.graph.current_node.meta["unbacked_bindings"]
torch._dynamo.exc.BackendCompilerFailed: backend='?' raised:
LoweringException: KeyError: 'unbacked_bindings'
  target: aten._local_scalar_dense.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg0_1', layout=FixedLayout('cpu', torch.float64, size=[], stride=[]))
  ))

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```

bobrenjc93 commented:
Sigh, it's related to [T205068920](https://www.internalfb.com/intern/tasks/?t=205068920)

Will re-land once https://www.internalfb.com/confighub/change/configo/4269739835 pushes

bobrenjc93 commented:
@pytorchbot merge

huydhn commented:
@bobrenjc93 I still see the same failure on the diff D65342565, please help take a look.  We would either need to come up with a fix quickly, or I need to revert it for a reland.  Keeping a diff for too long in the train opens up conflicts between github and fbcode
